### PR TITLE
Adds the notion of a local write set to contracts

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/assigns_enforce_18/test.desc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+--
+We had problems before when freeing a variable, but still kepting it on
+the writable set, which lead to dead variables issues. Anything that is local,
+including local allocations are now tracked by our freely assignable set;
+therefore, CBMC doesn't include these local variables in its alias check
+anymore (which avoids the issue previously described). Assignments to
+anything in our freely assignable set are bypassed. It’s important to
+notice that we track the address of all objects, thus, if a local pointer
+is created and points to something that is not local (e.g., inputs or global
+variables), CBMC will check every assignment to it.
```
Comment (TGWDB) [inline regression/contracts/assigns_enforce_18/test.desc]: "still kepting it on the writable set, which lead to dead variables issues." -> "still keeping it in the writable set, which lead to dead variable issues."

The information in this test appears to be useful to know, but in the wrong place. This describes how the implementation works, not what the test is actually checking for. I expect that someone trying to understand the code for contracts would look at the source code as a place to find information on how it works.

**Review by TGWDB — CHANGES_REQUESTED** [review]
Good to have a regression test (and the test looks good), please put the information from the test description comment close to where it would be found/used by a developer (to explain why the code works the way it does).

**Review by TGWDB — APPROVED** [review]

**Comment by feliperodri** [comment]
@SaswatPadhi could you take another look at it?

File: regression/contracts/assigns_enforce_18/test.desc [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+^\[foo.\d+\] line \d+ Check that a is assignable: SUCCESS$
+^\[foo.\d+\] line \d+ Check that yp is assignable: SUCCESS$
+^\[foo.\d+\] line \d+ Check that z is assignable: SUCCESS$
+^\[foo.\d+\] line \d+ Check that \*xp is assignable: SUCCESS$
+^\[foo.\d+\] line \d+ Check that y is assignable: FAILURE$
+^.* 1 of \d+ failed \(\d+ iteration.*\)$
+^VERIFICATION FAILED$
+--
+^\[foo.pointer\_primitives.\d+\] line \d+ deallocated dynamic object in POINTER_OBJECT\(tmp\_cc\$\d+\): FAILURE$
+^.* 8 of \d+ failed (\d+ iterations)$
```
Comment (SaswatPadhi) [inline regression/contracts/assigns_enforce_18/test.desc]: I don't understand this negative regex. Why are we checking `8` out of "anything"?
We should list out the ones of interest that are not supposed to pass / fail explicitly, just like the assertion above.

File: src/goto-instrument/contracts/assigns.h [hunk 3]
```diff
@@ -34,7 +34,7 @@ class assigns_clauset
 
     bool operator==(const targett &other) const
     {
-      return expr == other.expr;
+      return address == other.address;
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.h:37]: This is inconsistent with the hash function, because that hashes the `expr` not `address`.
We don't want to have two targets such that `t1 == t2` but `hash(t1) != hash(t2)`.

File: src/goto-instrument/contracts/assigns.cpp [hunk 4]
```diff
@@ -99,7 +99,8 @@ assigns_clauset::assigns_clauset(
 
 void assigns_clauset::add_target(const exprt &target_expr)
 {
-  auto insertion_succeeded = targets.emplace(*this, target_expr).second;
+  auto insertion_succeeded =
+    write_set.emplace(*this, targett::normalize(target_expr)).second;
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: I am not sure why this change is needed.
The constructor already calls `normalize` before storing the `address`.

I think this call should remain as before and we should be able to remove `expr` from `targett` entirely -- only `address` seems to be enough to track everything.

File: regression/contracts/assigns_validity_pointer_04/test.desc [hunk 5]
```diff
@@ -1,9 +1,11 @@
-KNOWNBUG
+CORE
 main.c
 --enforce-contract foo --replace-call-with-contract bar --replace-call-with-contract baz _ --pointer-primitive-check
-^EXIT=0$
+^EXIT=10$
 ^SIGNAL=0$
-^VERIFICATION SUCCESSFUL$
+^\[foo.\d+\] line \d+ Check that z is assignable: FAILURE$
+^.* 1 of \d+ failed \(\d+ iteration.*\)$
```
Comment (SaswatPadhi) [inline regression/contracts/assigns_validity_pointer_04/test.desc:7]: Same comment. Why are we checking `1` out of "anything"?

File: src/goto-instrument/contracts/assigns.cpp [hunk 6]
```diff
@@ -112,14 +113,32 @@ void assigns_clauset::add_target(const exprt &target_expr)
 
 void assigns_clauset::remove_target(const exprt &target_expr)
 {
-  targets.erase(targett(*this, targett::normalize(target_expr)));
+  write_set.erase(targett(*this, targett::normalize(target_expr)));
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: Same comment -- constructor calls `normalize` already. This should remain unchanged.

File: src/goto-instrument/contracts/assigns.cpp [hunk 7]
```diff
@@ -112,14 +113,32 @@ void assigns_clauset::add_target(const exprt &target_expr)
...
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
 }
 
 goto_programt assigns_clauset::generate_havoc_code() const
 {
   modifiest modifies;
-  for(const auto &target : targets)
-    modifies.insert(target.address.object());
+  for(const auto &target : write_set)
+    modifies.insert(to_address_of_expr(target.address).object());
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: But `address` is already an `address_of_exprt`

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
I am leaving some initial comments below, but I think we should discuss these two sets that are being maintained within assigns clause now. I think an assigns clause should just refer to the assigns clause on a function / loop -- it shouldn't be modified with other stuff.

We could create a separate class, may be called `assignable_set`, which would represent a union of the assigns clause + freely assignable local symbols. Let's discuss more offline.

File: src/goto-instrument/contracts/assigns.h (hunk 3 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.h:37]: You're correct. I fixed the hash function as well.

File: src/goto-instrument/contracts/assigns.cpp (hunk 4 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Correct, we don't need to normalize this here again and we can completely remove `expr`. Done.

File: src/goto-instrument/contracts/assigns.cpp (hunk 6 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Correct, remove it.

File: src/goto-instrument/contracts/assigns.cpp (hunk 7 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Correct. This is unnecessary. Removed it.

File: regression/contracts/assigns_validity_pointer_04/test.desc (hunk 5 above)
Comment (SaswatPadhi) [reply regression/contracts/assigns_validity_pointer_04/test.desc:7]: Thanks for explaining offline. Resolving this.

File: src/goto-instrument/contracts/assigns.cpp [hunk 8]
```diff
@@ -112,13 +111,30 @@ void assigns_clauset::add_target(const exprt &target_expr)
...
+  {
+    log.warning() << "Ignored duplicate expression '"
+                  << from_expr(ns, expr.id(), expr) << "' in assigns clause at "
+                  << expr.source_location().as_string() << messaget::eom;
+  }
+}
+
+void assigns_clauset::remove_freely_assignable_expr(const exprt &expr)
+{
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]:
```suggestion
  freely_assignable_set.erase(targett(*this, expr));
```
since `normalize` is called by the constructor.

File: src/goto-instrument/contracts/assigns.cpp [hunk 9]
```diff
@@ -145,11 +166,11 @@ exprt assigns_clauset::generate_containment_check(const exprt &lhs) const
 exprt assigns_clauset::generate_subset_check(
   const assigns_clauset &subassigns) const
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp:164]: This function needs to be updated for our "two sets" scheme now. We are only checking that:
`subassigns.write_set ⊆ write_set`.

I think we should be checking that:
`(subassigns.freely_assignable_set = {}) ∧ (subassigns.write_set ⊆ write_set ∪ freely_assignable_set)`

File: src/goto-instrument/contracts/assigns.cpp [hunk 10]
```diff
@@ -112,13 +111,30 @@ void assigns_clauset::add_target(const exprt &target_expr)
...
+{
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
 }
 
 goto_programt assigns_clauset::generate_havoc_code() const
 {
   modifiest modifies;
-  for(const auto &target : targets)
+  for(const auto &target : write_set)
     modifies.insert(target.address.object());
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp:135]: For completeness, we should also havoc the `freely_assignable_set`. So we need a similar loop for that.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Some more comments.

File: regression/contracts/assigns_validity_pointer_04/test.desc (hunk 5 above)
Comment (feliperodri) [reply regression/contracts/assigns_validity_pointer_04/test.desc:7]: For this case, it's OK to track `1` of "anything". Pointer primitive check can generate many properties to check and I want to make sure that there is only 1 failure: `Check that z is assignable`.

File: src/goto-instrument/contracts/assigns.cpp (hunk 8 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Done.

File: src/goto-instrument/contracts/assigns.cpp (hunk 10 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp:135]: Implemented as suggested.

File: src/goto-instrument/contracts/assigns.cpp (hunk 9 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp:164]: Implemented as suggested.

File: regression/contracts/assigns_enforce_18/test.desc (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/assigns_enforce_18/test.desc]: I ran this regression test with our `develop` branch and included the failed checks in the negative regex.

File: src/goto-instrument/contracts/assigns.cpp [hunk 11]
```diff
@@ -89,17 +88,18 @@ assigns_clauset::assigns_clauset(
...
 {
   for(const auto &target_expr : expr.operands())
   {
-    add_target(target_expr);
+    add_global_write_set(target_expr);
   }
 }
 
-void assigns_clauset::add_target(const exprt &target_expr)
+void assigns_clauset::add_global_write_set(const exprt &target_expr)
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: [nitpick]

I liked the names we discussed, `add_to_global_write_set` and `remove_from_global_write_set` better. The current names are a bit confusing because it sounds like we are adding a whole set or removing a whole set.

Before it was `add_target` and that made sense, because we were adding a new target.

File: src/goto-instrument/contracts/assigns.cpp [hunk 12]
```diff
@@ -145,11 +170,15 @@ exprt assigns_clauset::generate_containment_check(const exprt &lhs) const
 exprt assigns_clauset::generate_subset_check(
   const assigns_clauset &subassigns) const
 {
-  if(subassigns.targets.empty())
+  if(subassigns.global_write_set.empty())
     return true_exprt();
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp:168]: We should add an `INVARIANT` at this point to make sure we don't internally call this function on a `subassigns` that has a non-empty `local_write_set`, because `subassigns.local_write_set` is entirely ignored in the rest of the function -- we actually don't expect it for the callee's assigns clause.

File: src/goto-instrument/contracts/assigns.cpp [hunk 13]
```diff
@@ -145,11 +170,15 @@ exprt assigns_clauset::generate_containment_check(const exprt &lhs) const
...
 
+  std::unordered_set<targett, targett::hasht> caller_write_set(
+    global_write_set);
+  caller_write_set.insert(local_write_set.begin(), local_write_set.end());
+
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp:173]: I think it might be cleaner (and more efficient) to just have another loop. The loop only has a single statement in it anyway. What do you think?

File: regression/contracts/assigns_enforce_19/main.c [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  a++;
+  return a;
+}
+
+int g(int *points_to_b, int *points_to_c)
+  __CPROVER_assigns(b) // Error: it should also mention c
+{
+  b = 1;
+  c = 2;
+  *points_to_b = 1;
```
Comment (SaswatPadhi) [inline regression/contracts/assigns_enforce_19/main.c:18]: Nice test case! We are now correctly looking at assignable (and freely assignable) "addresses" instead of symbols :)

File: src/goto-instrument/contracts/assigns.cpp [hunk 15]
```diff
@@ -110,32 +110,57 @@ void assigns_clauset::add_target(const exprt &target_expr)
...
+}
+
+void assigns_clauset::add_local_write_set(const exprt &expr)
+{
+  auto insertion_succeeded = local_write_set.emplace(*this, expr).second;
+
+  if(!insertion_succeeded)
+  {
+    log.warning() << "Ignored duplicate expression '"
+                  << from_expr(ns, expr.id(), expr) << "' in assigns clause at "
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: This warning message can be a little confusing, since the user-specified targets only appear within `global_write_set`.

Should we just remove this warning and silently drop any local targets that we might internally add accidentally? Or, we could fatally fail on an `INVARIANT` failure here if we want to be absolutely sure that we never double-add local targets, but I think that might be too aggressive.

**Review by SaswatPadhi — COMMENTED** [review]
Thanks! :) A few last comments

File: src/goto-instrument/contracts/assigns.cpp (hunk 11 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Implemented as suggested.

File: src/goto-instrument/contracts/assigns.cpp (hunk 12 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp:168]: Implemented as suggested.

File: src/goto-instrument/contracts/assigns.cpp (hunk 15 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Implemented as suggested.

File: src/goto-instrument/contracts/assigns.cpp (hunk 13 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp:173]: Implemented as suggested.

File: regression/contracts/assigns_enforce_18/test.desc [hunk 16]
```diff
@@ -0,0 +1,22 @@
+CORE
+main.c
+--enforce-all-contracts _ --pointer-primitive-check
+^EXIT=10$
+^SIGNAL=0$
+^\[bar.\d+\] line \d+ Check that \*b is assignable: SUCCESS$
```
Comment (SaswatPadhi) [inline regression/contracts/assigns_enforce_18/test.desc:7]: We don't check that `a` is assignable at `free(a)`, or may be we do but don't have a regex for it?

File: regression/contracts/assigns_validity_pointer_04/test.desc (hunk 5 above)
Comment (SaswatPadhi) [reply regression/contracts/assigns_validity_pointer_04/test.desc:7]: I am a little confused as to why only 1 assertion fails here though.

The assigns clause for `foo` doesn't have `z`, nor does it have `*z`. So shouldn't we fail 2 assertions:
1. the containment check @ assignment `z = malloc...`
2. the subset check @ call `baz()` in the next line (which has `*z` in the assign clause for `baz` but not `foo`)?

**Review by SaswatPadhi — APPROVED** [review]
Awesome 🎉

2 comments on the regression tests, but otherwise LGTM.

File: regression/contracts/assigns_validity_pointer_04/test.desc (hunk 5 above)
Comment (feliperodri) [reply regression/contracts/assigns_validity_pointer_04/test.desc:7]: We discussed this offline. Two important points:
1. There is no containment check here, `malloc` is not treated as a normal function call.
2. There is a bug in checking the subset relation which is not stressed by this test case. We will address it on a separate PR.

File: regression/contracts/assigns_enforce_18/test.desc (hunk 16 above)
Comment (feliperodri) [reply regression/contracts/assigns_enforce_18/test.desc:7]: Good catch!! I just sent a new commit to check frame conditions on free function calls.

File: regression/contracts/assigns_validity_pointer_04/test.desc (hunk 5 above)
Comment (SaswatPadhi) [reply regression/contracts/assigns_validity_pointer_04/test.desc:7]: > There is no containment check here, `malloc` is not treated as a normal function call.

Hmm, but `malloc` just returns a fresh allocation, the left-hand side of the assignment in `z = malloc...`, i.e. the variable `z`, must be checked for containment in the assigns clause.

> There is a bug in checking the subset relation which is not stressed by this test case. We will address it on a separate PR.

On digging a little bit deeper, I realized that the we don't actually need to separately check the assigns clause during replacement. We can go ahead and havoc directly, and those havocs would fail if something is not assignable in the calling context.
